bpf-firewall: custom BPF programs through IP(Ingress|Egress)FilterPath=

pothos opened the pull request:
Takes a single `/sys/fs/bpf/pinned_prog` string as argument but may be
specified multiple times. An empty assignment resets all previous filters.

Closes https://github.com/systemd/systemd/issues/10227

I propose to use filter paths and not a single boolean with a predefined path to lookup the pinned BPF program. With the chosen approach of this PR multiple filters can be combined, e.g., to attach a filter for HTTP contents and a separate filter for certain IP addresses. It also allows to reuse filters between units where, e.g., multiple units could depend on a oneshot unit that loads the filter at a certain place.

I decided for absolute paths since it makes reading and using these values elsewhere easy. It is more clear for users to see what should be pinned where in order to work (and other directives do also not strip away `/dev/` in `/dev/sda1`).

Does this make sense? I can change it to use relative paths etc. Did I forget something such as cleanup macros?
Currently, as with `IPAddressDeny` it is not a failure if the BPF program cannot be attached or `bpf_program_new` fails. The error is logged but the unit will still start. Should this be more strict since the `IPAddressDeny` options are defaults somewhere but this one here not?

What about weight 500 in the security analysis, is it too high? It's not recommended to have a custom filter but having it present may indicate a stronger firewalling. If egress, ingress or both are firewalled depends on the situation, so there is varying score, just on or off.

pothos commented:
One more note about boolean arguments: How would they work with `systemd-run`?

poettering commented on src/analyze/analyze-security.c:
```diff
@@ -613,6 +616,43 @@ static int assess_ip_address_allow(
+        } else {
+                d = strdup("Service defines a custom egress IP filter with a BPF program");
+                b = 0;
+        }
```
hmm, this should be tweaked I think. If a service has IPAddressAllow= defined, then not having IPFilterIngress= shouldn't be considered "bad". Similar, if IPAddressAllow= not being set shouldn't count as bad if IPFilterIngress= is set. i.e. the two tests should check for each other (but should be kept as two entries in the output). This requires a bit of care though, as if neither is set, we don't want both to account for the badness...

poettering commented on src/analyze/analyze-security.c:
```diff
@@ -1824,6 +1871,44 @@ static int property_read_ip_address_allow(
+        for (;;) {
+                const char *path;
+
+                r = sd_bus_message_read(m, "i", &path);
```
this can't be right. you are reading an integer here. Needs to be a string

pothos replied:
Good catch, thanks.

poettering commented on src/analyze/analyze-security.c:
```diff
@@ -1824,6 +1871,44 @@ static int property_read_ip_address_allow(
+                        break;
+
+                if (streq(member, "IPFilterIngress")) {
+                        info->ip_filters_custom_ingress = strlen(path) > 0;
```
use "isempty(path)", it's a lot faster (since it runs in O(1) time, not O(n)).

poettering commented on src/shared/bus-unit-util.c:
```diff
@@ -716,6 +716,19 @@ static int bus_append_cgroup_property(sd_bus_message *m, const char *field, cons
+                        r = sd_bus_message_append(m, "(sv)", field, "as", 0);
+                else {
+                        r = sd_bus_message_append(m, "(sv)", field, "as", 1, eq);
+                }
```
no {} around single line if blocks, see https://systemd.io/CODING_STYLE#formatting

poettering commented on src/core/load-fragment.c:
```diff
@@ -4457,6 +4458,76 @@ int config_parse_disable_controllers(
+        STRV_FOREACH(t, *paths) {
+                if (path_equal(resolved, *t))
+                        return 0;
+        }
```
there should be a helper path_strv_contains() that does this, similar to the existing strv_contains(). Should be in path-util.c I guess, next to the definition of path_equal()

pothos replied:
I think `strv_contains` does the same thing in this case, so I just used it.

poettering commented on src/core/load-fragment.c:
```diff
@@ -4457,6 +4458,76 @@ int config_parse_disable_controllers(
+        r = strv_extend(paths, resolved);
+        if (r == -ENOMEM) {
+                return log_oom();
+        }
```
no {} around single-line if blocks...

also, you should catch all failures of strv_extend() here, not just ENOMEM. That said, the call only really fails on ENOMEM hence it please just write this as:

```c
r = strv_extend(…);
if (r < 0)
        return log_oom();
```

poettering commented on src/core/dbus-cgroup.c:
```diff
@@ -1441,6 +1468,100 @@ int bus_cgroup_set_property(
+                                                *filters = new0(char*, 1);
+                                                if (!*filters)
+                                                       return log_oom();
+                                        }
```
this is unnecessary, strv_extend() will do the right thing when the strv is NULL anyway...

poettering commented on src/core/dbus-cgroup.c:
```diff
@@ -1441,6 +1468,100 @@ int bus_cgroup_set_property(
+                                                skip_entry = true;
+                                                break;
+                                        }
+                                }
```
please add path_strv_contains() as common helpe to path-util.c (see other comment)

poettering commented on src/core/dbus-cgroup.c:
```diff
@@ -1441,6 +1468,100 @@ int bus_cgroup_set_property(
+                                        r = strv_extend(filters, path);
+                                        if (r == -ENOMEM) {
+                                                return log_oom();
+                                        }
```
```c
if (r < 0)
        return log_oom();
```

poettering commented on src/core/dbus-cgroup.c:
```diff
@@ -1441,6 +1468,100 @@ int bus_cgroup_set_property(
+
+                        STRV_FOREACH(entry, *filters) {
+                                fprintf(f, "%s=%s\n", name, *entry);
+                        }
```
no {} around single line if blocks

poettering commented on src/core/load-fragment.c:
```diff
@@ -4457,6 +4458,76 @@ int config_parse_disable_controllers(
+                *paths = new0(char*, 1);
+                if (!*paths)
+                        return log_oom();
+        }
```
unnecessary, strv_extend() does the right thing anyway, if the first param points to a  NULL ptr

poettering commented on src/shared/bpf-program.c:
```diff
@@ -235,3 +254,47 @@ int bpf_map_lookup_element(int fd, const void *key, void *value) {
+        return 0;
+}
+
+
```
spurious double newline

poettering commented on src/core/dbus-cgroup.c:
```diff
@@ -1441,6 +1468,100 @@ int bus_cgroup_set_property(
+                const char *path;
+                char **t;
+                char **entry;
+                char ***filters;
```
please keep variable declarations always at the innermost scope

poettering commented on src/core/dbus-cgroup.c:
```diff
@@ -320,6 +320,31 @@ static int property_get_ip_address_access(
+        char **entry;
+        int r;
+
+        r = sd_bus_message_open_container(reply, 'a', "s");
```
use sd_bus_message_append_strv() instead, it does the same thing in one go. That said, you can actually drop the whole function, since sd-bus does such a serialization anyway implicitly if it encounters an "as" property in the vtable, and the getter is set to NULL. Hence, please use that, drop this getter routine and just pass NULL. It works the same with zero code.

poettering commented on src/core/cgroup.c:
```diff
@@ -360,6 +361,14 @@ void cgroup_context_dump(CGroupContext *c, FILE* f, const char *prefix) {
+
+        STRV_FOREACH(path, c->ip_filters_ingress) {
+                fprintf(f, "%sIPFilterIngress=%s/\n", prefix, *path);
+        }
```
no {} around single-line blocks...

Hmm, why the / suffix? are programs in bpffs exposed as directories?

poettering commented on src/shared/bpf-program.h:
```diff
@@ -40,3 +41,24 @@ int bpf_map_update_element(int fd, const void *key, void *value);
+struct BPFProgramList {
+        unsigned n_ref;
+        LIST_HEAD(BPFProgramItem, list);
+};
```
hmm, why do we need two new types for that? i mean, if we want to allow multiple of these then it might be simpler to just use a `Set`. Or maybe even a realloc()able array, after all this is not going to be that many new things.

Moreover, it appears to me that having this ref-counted as a whole is unnecessary. It makes more sense to me, to ref count only the programs themselves as we do, but inc their ref as soon as we add them two any `Set`... I mean, the activation of all bpf progs is not going to be atomic anyway but step-by-step...

poettering commented on src/core/unit.h:
```diff
@@ -281,6 +281,10 @@ typedef struct Unit {
+        BPFProgramList *ip_bpf_custom_ingress;
+        BPFProgramList *ip_bpf_custom_ingress_installed;
+        BPFProgramList *ip_bpf_custom_egress;
+        BPFProgramList *ip_bpf_custom_egress_installed;
```
should really bet `Set*` objects I am sure, see other comment

pothos replied:
Thanks :)

poettering commented on src/core/load-fragment-gperf.gperf.m4:
```diff
@@ -202,6 +202,8 @@ $1.DisableControllers,           config_parse_disable_controllers,   0,
 $1.IPAddressAllow,               config_parse_ip_address_access,     0,                             offsetof($1, cgroup_context.ip_address_allow)
 $1.IPAddressDeny,                config_parse_ip_address_access,     0,                             offsetof($1, cgroup_context.ip_address_deny)
+$1.IPFilterIngress,              config_parse_ip_filter_bpf_progs,   0,                             offsetof($1, cgroup_context.ip_filters_ingress)
+$1.IPFilterEgress,               config_parse_ip_filter_bpf_progs,   0,                             offsetof($1, cgroup_context.ip_filters_egress)
```
So, I'd put more focus on the fact we specify programs by fs path here. i.e. maybe name this IPIngressFilterPath= or so. This would then nicely match NetworkNamespacePath= which semantically is very similar I think. I.e. what differentiates this new option from IPAddressAllow= and IPAddressDeny= is not whether its a filter or not (because all of them are), but that we specify the filter by path in fs here.

pothos replied:
Yes, changed!

poettering commented on src/core/bpf-firewall.c:
```diff
@@ -539,6 +543,52 @@ int bpf_firewall_compile(Unit *u) {
+                        }
+                        LIST_APPEND(items, u->ip_bpf_custom_egress->list, prog_item);
+                        prog_item = NULL;
+                }
```
this looks like like these two loops could be unified in one, if they split out into a function that is called twice, once with ip_filters_ingress and once with ip_filters_egress...

poettering commented:
> Should this be more strict since the `IPAddressDeny` options are defaults somewhere but this one here not?

I think so, yes. I mean, the reason the IPAddressAllow= and IPAddressDeny= lines are gracefully skipped is that we want people to use them by default in upstream packaging (for example to limit traffic to localhost), and things shall still work when these unit files are used in environments where bpf is not available (because the kernel doesn't do it, or running in a container or such). But since these kind of manual bpf programs don't make much sense in similar upstream scenarios, since they always need manual setup, i would say these ones should be fatal by default, if they can't be installed.

pothos commented:
Thank you for the review and good hints! I have addressed the issues and think it is in a better state now.

Actually I removed the entry in `systemd-analyze security` because the new logic would mean that the badness is always off. But I took the existance of custom filters into account in the `IPAddressDeny` test. 

I wanted to introduce error handling for `cgroup_context_apply` (with a definition for fatal errors to not change behavior) and propagate it to `unit_realize_cgroup`. What do you think? Since it looks like a bigger change, I decided to just work around it  with a small test to load the BPF programs in `unit_start` to fail the start if the paths are not valid.

poettering commented on man/systemd.resource-control.xml:
```diff
@@ -614,6 +614,39 @@
+        <listitem>
+          <para>Add custom network traffic filters implemented as BPF programs, applying to all IP packets
+          sent and received over <constant>AF_INET</constant> and <constant>AF_INET6</constant> sockets.
+          Takes an absolute path to a pinned BPF program in the BPF virtual filesystem.
```
maybe say `BPF virtual filesystem (<filename>/sys/fs/bpf/</filename>)`, i.e. indicate that it's mounted there.

poettering commented on man/systemd.resource-control.xml:
```diff
@@ -614,6 +614,39 @@
+          <varname>IPAddressAllow=</varname> and <varname>IPAddressDeny=</varname> filters in any of these units.
+          By default there are no filters specified.</para>
+
+          <para>If these settings are used multiple times in the same unit the specified programs are all attached. If an
```
the specified programs are all → al the specified programs are

poettering commented on man/systemd.resource-control.xml:
```diff
@@ -614,6 +614,39 @@
+          <para>If these settings are used multiple times in the same unit the specified programs are all attached. If an
+          empty string is assigned to these settings the program list is reset and all previous specified programs ignored.</para>
+
+          <para>Note that for socket-activated services, the IP access list configured on the socket unit applies to
```
`IP Access list (as configured with IPAddressAllow=/IPAddressDeny= above)`

pothos replied:
Sorry, this was a copy error, I wanted to write about the filter programs which have to be configured for both the socket unit and the service unit, not the access list with allow/deny.

poettering commented on src/analyze/analyze-security.c:
```diff
@@ -1824,6 +1830,44 @@ static int property_read_ip_address_allow(
+        assert(bus);
+        assert(member);
+        assert(m);
+
```
please rewrite as:

```c
r = sd_bus_message_read_strv(m, &l);
if (r < 0)
        return r;
if (streq(member, "IPIngressFilterPath"))
        info->ip_filters_custom_ingress = !strv_isempty(l);
else if (streq(…
```

pothos replied:
Thanks!

poettering commented on src/core/bpf-firewall.c:
```diff
@@ -542,11 +542,73 @@ int bpf_firewall_compile(Unit *u) {
 
+static int load_bpf_progs_from_fs_to_set(Unit *u, char **filter_paths, Set **set) {
+        char **bpf_fs_path;
+        _cleanup_free_ BPFProgram *prog = NULL;
```
please move this to the innermost scope this is used

poettering commented on src/core/bpf-firewall.c:
```diff
@@ -542,11 +542,73 @@ int bpf_firewall_compile(Unit *u) {
+        _cleanup_free_ BPFProgram *prog = NULL;
+        int r;
+
+        *set = set_free_with_destructor(*set, bpf_program_unref);
```
for new code, please use a "hashops" structure when doing set_new() that has a desctructor call set. With that you can then just do set_free() and the destructor is implicitly called. In fact, with that you can even do set_clear() here, which is a bit nicer I think.

pothos replied:
Yes, that's better :)

poettering commented on src/core/bpf-firewall.c:
```diff
@@ -542,11 +542,73 @@ int bpf_firewall_compile(Unit *u) {
+
+        if (supported != BPF_FIREWALL_SUPPORTED_WITH_MULTI) {
+                log_unit_debug(u, "BPF_F_ALLOW_MULTI not supported on this manager, cannot attach custom BPF programs.");
+                return -EOPNOTSUPP;
```
in new code, please write this as

```c
if (supported != …)
        return log_unit_debug_errno(u, SYNTHETIC_ERRNO(EOPNOTSUPP), "…");
```

poettering commented on src/core/bpf-firewall.c:
```diff
@@ -592,13 +662,42 @@ int bpf_firewall_install(Unit *u) {
+                if (r < 0)
+                        return log_unit_error_errno(u, r, "Can't allocate BPF program set: %m");
+
+                r = set_put(u->ip_bpf_custom_egress_installed, bpf_program_ref(prog));
```
if set_put() fails, you leave a reference dangling, and thus leak the bpf program...

poettering commented on src/core/bpf-firewall.c:
```diff
@@ -592,13 +662,42 @@ int bpf_firewall_install(Unit *u) {
+                r = set_put(u->ip_bpf_custom_ingress_installed, bpf_program_ref(prog));
+                if (r < 0)
+                        return log_unit_error_errno(u, r, "Can't add program to BPF program set: %m");
+        }
```
it appears to me it would be relatively nice and easy to place this loop into a function that an then be called once for ingress and once for egress

poettering commented on src/core/unit.c:
```diff
@@ -1773,6 +1779,11 @@ int unit_start(Unit *u) {
+        /* Load any custom firewall BPF programs here once to test if they are existing and actually loadable.
+         * Fail here early since later errors in the call chain unit_realize_cgroup to cgroup_context_apply are ignored. */
+        if (unit_get_cgroup_context(u) && bpf_firewall_load_custom(u))
+                return log_unit_notice_errno(u, SYNTHETIC_ERRNO(EPROTO), "Loading custom BPF programs for firewall failed.");
```
please do this in unit_prepare_exec() so that this is only done when we execute stuff, not ahead of time.

pothos replied:
I tried this but it does not work for `systemd-run --scope`, so I would leave it where it is - or any other place?

poettering replied:
hmm, I think adding it to unit_attach_pids_to_cgroup() might make sense, as that is called in the scope case at the moment we add the first PIDs to the scope. If this is called from both unit_prepare_exec() and unit_attach_pids_to_cgroup() then we'd cover every case where we are about to add PIDs to a cgroup.

Hmm, and  I think it would make sense to make bpf_firewall_load_custom() simply return 0 if unit_get_cgroup_context() doesn't hold. i.e. if a unit doesn't have cgroups, it can't have custom bpf filters either, and we can definitly load "nothing" successfully. That should simplify this a bit.

Also, always propagate the original error. Here bpf_firewall_load_custrom( ) returns a proper error code, but you convert that to EPROTO for no reason...

also, I think bpf_firewall_load_custrom() should log about all errors on its own. that would make the invocation here much simpler, as you don't have to generate a msg here at all..

(background: in systemd we want that functions either log about all errors or about none. In the latter case the caller should log. Hence, whenever you write a function figure out if it shall be a logging or no-logging one, and then stick to it. Logging at LOG_DEBUG level is not considered logging in this case btw, everybody is allowed to do that)

pothos commented:
Have addressed the issues (also ASAN), except moving the check from `unit_start` to `unit_prepare_exec` as mentioned in the review comment.

evverx commented:
@pothos as it turns out, I was too optimistic about how easy it is to run test-bpf on Travis CI under ASan/UBsan :-) I hope @mrc0mmand will help to bring them to Arch (run on CentOS CI with vagrant) so that it could be covered. In the meantime, I ran the test under Valgrind manually and the memory leak seems to be gone.

By the way, there was an attempt to start using libbpf in https://github.com/systemd/systemd/pull/12151. I'm wondering if it would be possible to somehow sync these two PRs. cc @wat-ze-hex @rdna.

rdna commented:
@evverx this PR could save a few lines of code if libbpf was in systemd, specifically it has both `bpf_obj_get` and `bpf_obj_pin` (see https://github.com/libbpf/libbpf/blob/master/src/bpf.c). 

Admittedly not too many though so IMO it should not be blocked by #12151 and it'll be easier to follow up later to switch to those libbpf functions if libbpf support is merged.

poettering commented on src/analyze/analyze-security.c:
```diff
@@ -1824,6 +1830,33 @@ static int property_read_ip_address_allow(
+                void *userdata) {
+
+        struct security_info *info = userdata;
+        char **l;
```
needs to be freed. sd_bus_message_read_strv() allocates...

i.e. change this to:

```c
_cleanup_(strv_freep) char **l = NULL;
```

pothos replied:
Thanks

pothos commented:
Hi,
thanks for the review and sorry for the big delay. I've moved the call and confirmed that `systemd-run` behaves as expected if given a wrong path, both with and without `--scope`.

poettering commented on src/core/bpf-firewall.c:
```diff
@@ -542,6 +542,95 @@ int bpf_firewall_compile(Unit *u) {
+
+                r = set_put(*set_installed, bpf_program_ref(prog));
+                if (r < 0) {
+                        bpf_program_unref(prog);
```
usually, we'd only add the ref after the set_put() success (so that we don't need the unref on error), but this works too

pothos replied:
Ok, I moved it into the else branch to make more clear that the ref action belongs to the result check.

poettering commented on src/core/bpf-firewall.c:
```diff
@@ -583,7 +675,8 @@ int bpf_firewall_install(Unit *u) {
         if (u->ip_bpf_egress) {
-                r = bpf_program_cgroup_attach(u->ip_bpf_egress, BPF_CGROUP_INET_EGRESS, path, flags);
+                r = bpf_program_cgroup_attach(u->ip_bpf_egress, BPF_CGROUP_INET_EGRESS, path,
+                                              !set_isempty(u->ip_bpf_custom_egress) ? BPF_F_ALLOW_MULTI : flags);
```
it doesn't really matter, but I think it would be nicer to do:

```
…, flags | (set_isempty(u->ip_bpf_custom_egress) ? 0 : BPF_F_ALLOW_MULTI));
```

i.e. always OR in the `flags` field, even though sometimes it's redundant. That way, maybe one day if a new flags value is defined we don't lose it here.

pothos replied:
Good point!

poettering commented on src/core/bpf-firewall.c:
```diff
@@ -592,13 +685,22 @@ int bpf_firewall_install(Unit *u) {
         if (u->ip_bpf_ingress) {
-                r = bpf_program_cgroup_attach(u->ip_bpf_ingress, BPF_CGROUP_INET_INGRESS, path, flags);
+                r = bpf_program_cgroup_attach(u->ip_bpf_ingress, BPF_CGROUP_INET_INGRESS, path,
+                                              !set_isempty(u->ip_bpf_custom_ingress) ? BPF_F_ALLOW_MULTI : flags);
```
as above

poettering commented on src/test/test-bpf.c:
```diff
@@ -88,14 +92,31 @@ int main(int argc, char *argv[]) {
+                        .file_flags = 0,
+                };
+
+                unlink(test_prog);
```
please cast this to `(void)`. We usually do that when we actively want to ignore a return value, as a hint to static checkers that we knowingly ignore return codes.

pothos replied:
sure

poettering commented on src/test/test-bpf.c:
```diff
@@ -88,14 +92,31 @@ int main(int argc, char *argv[]) {
+                unlink(test_prog);
+
+                r = bpf(BPF_OBJ_PIN, &attr, sizeof(attr));
+                if (r) {
```
we usually check for syscall failure with a `< 0` check, please do that here too

pothos replied:
Ok, done :)

poettering commented on src/test/test-bpf.c:
```diff
@@ -88,14 +92,31 @@ int main(int argc, char *argv[]) {
+
+                r = bpf(BPF_OBJ_PIN, &attr, sizeof(attr));
+                if (r) {
+                        log_notice("BPF object pinning failed, will not run custom filter test.");
```
please include the error string here, i.e:

```c
log_warning_errno(errno, "BPF object …: %m");
```

pothos replied:
Thanks

poettering commented on src/test/test-bpf.c:
```diff
@@ -175,5 +196,31 @@ int main(int argc, char *argv[]) {
+                assert_se(SERVICE(u)->exec_command[SERVICE_EXEC_START]->exec_status.code != CLD_EXITED ||
+                          SERVICE(u)->exec_command[SERVICE_EXEC_START]->exec_status.status != EXIT_SUCCESS);
+
+                unlink(test_prog);
```
cast to (void) please

pothos commented:
@poettering Have addressed it, thanks for bundling the review with actual code suggestions :+1:

poettering commented on src/core/bpf-firewall.c:
```diff
@@ -539,6 +539,95 @@ int bpf_firewall_compile(Unit *u) {
+                if (r < 0)
+                        return log_unit_error_errno(u, r, "Can't allocate BPF program set: %m");
+
+                r = set_put(*set, TAKE_PTR(prog));
```
if set_put() fails, then this already invalidated prog and thus the memory prog points to is leaked. Please rewrite as:

```c
r = set_put(*set, prog);
if (r < 0)
        return …;
TAKE_PTR(prog);
```

so that `prog` is only set to NULL after we are sure it's in the set.

pothos replied:
Ah, yes, tricky…

poettering commented on src/core/bpf-firewall.c:
```diff
@@ -539,6 +539,95 @@ int bpf_firewall_compile(Unit *u) {
+                r = set_put(*set_installed, prog);
+                if (r < 0)
+                        return log_unit_error_errno(u, r, "Can't add program to BPF program set: %m");
+                else
```
no need for the else here... in the interest to keeping the indentation level small, let's just remove it

pothos commented:
Thanks, done (:

poettering commented:
Excellent! Thanks!

pothos commented:
Actually forgot to add the `NEWS` entry to the commit :see_no_evil:
